drivers/at25xxx/at25xxx.c

```diff
@@ -0,0 +1,206 @@
+#ifdef USEMODULE_XTIMER
+        xtimer_usleep(POLL_DELAY_US);
+#endif
+    }
+
+    /* set write enable and wait for status change */
+    spi_transfer_byte(dev->params.spi, dev->params.cs_pin, false, CMD_WREN);
+    while (!_write_enabled(dev)) {}
+
+    /* write the data */
+    spi_transfer_bytes(dev->params.spi, dev->params.cs_pin, true, &pos, NULL, 1 + ADDR_LEN / 8);
+    spi_transfer_bytes(dev->params.spi, dev->params.cs_pin, false, data, NULL, len);
+
+    return len;
+}
+
+size_t at25xxx_write(const at25xxx_t *dev, uint32_t pos, const void *data, size_t len)
+{
+    const uint8_t *d = data;
+
+    if (pos + len > dev->params.size) {
+        errno = ERANGE;
+        return 0;
+    }
+
+    getbus(dev);
+
+    while (len) {
+        size_t written = _write_page(dev, pos, d, len);
+        len -= written;
+        pos += written;
+        d   += written;
+    }
+
+    spi_release(dev->params.spi);
+
+    return (void*)d - data;
+}
+
+void at25xxx_write_byte(const at25xxx_t *dev, uint32_t pos, uint8_t data)
+{
+    at25xxx_write(dev, pos, &data, sizeof(data));
+}
+
+size_t at25xxx_read(const at25xxx_t *dev, uint32_t pos, void *data, size_t len)
+{
+    if (pos + len > dev->params.size) {
+        errno = ERANGE;
+        return 0;
+    }
+
+    getbus(dev);
+
+    /* wait for previous write to finish - may take up to 5 ms */
+    while (_write_in_progress(dev)) {
+#ifdef USEMODULE_XTIMER
+        xtimer_usleep(POLL_DELAY_US);
+#endif
+    }
+
+    pos = _pos(CMD_READ, pos);
+    spi_transfer_bytes(dev->params.spi, dev->params.cs_pin, true, &pos, NULL, 1 + ADDR_LEN / 8);
+    spi_transfer_bytes(dev->params.spi, dev->params.cs_pin, false, NULL, data, len);
+
+    spi_release(dev->params.spi);
+
+    return len;
+}
+
+uint8_t at25xxx_read_byte(const at25xxx_t *dev, uint32_t pos)
+{
+    uint8_t b;
+    at25xxx_read(dev, pos, &b, sizeof(b));
+    return b;
+}
+
+size_t at25xxx_set(const at25xxx_t *dev, uint32_t pos, uint8_t val, size_t len)
+{
+    uint8_t data[PAGE_SIZE];
+    size_t total = 0;
+
+    if (pos + len > dev->params.size) {
+        errno = ERANGE;
+        return 0;
+    }
+
+    memset(data, val, sizeof(data));
+
+    getbus(dev);
+
+    while (len) {
+        size_t written = _write_page(dev, pos, data, len);
+        len   -= written;
+        pos   += written;
+        total += written;
+    }
+
+    spi_release(dev->params.spi);
+
+    return total;
+}
+
+size_t at25xxx_clear(const at25xxx_t *dev, uint32_t pos, size_t len)
+{
+    return at25xxx_set(dev, pos, 0, len);
+}
+
+void at25xxx_init(at25xxx_t *dev, const at25xxx_params_t *params)
+{
+    dev->params = *params;
+    spi_init_cs(dev->params.spi, dev->params.cs_pin);
+
+    if (dev->params.wp_pin != GPIO_UNDEF) {
+        gpio_init(dev->params.wp_pin, GPIO_OUT);
+        gpio_set(dev->params.wp_pin);
+    }
+
+    if (dev->params.hold_pin != GPIO_UNDEF) {
+        gpio_init(dev->params.hold_pin, GPIO_OUT);
+        gpio_set(dev->params.hold_pin);
+    }
```

Comment:
`spi_acquire()` (called in `get_bus()`) can fail. There are two schools of thought. One is to always check the return value of `spi_acquire()`, the other is to only check it during startup. (The reasoning for only checking during startup is, that SPI acquire can only legitimately fail if it is called with invalid parameters of an incompatible configuration. As the parameters and the configuration are not changed, checking the return value once is sufficient.)

Normally RIOT drivers are required to  [verify connectivity and reset the device](https://github.com/RIOT-OS/RIOT/wiki/Guide%3A-Writing-a-device-driver-in-RIOT#initialization) during initialization. I only briefly skimmed the data sheet and could not find a way to reset the device (except for the power-on-reset in 5.1.2) nor could I find some register providing a well-known constant device ID or similar to confirm connectivity. But checking if `spi_acquire()` succeeds in order to validate that the SPI config is compatible with the board is possible and should be done.

Reply:
Ok, will do!